src/ipc/ipc4/handler.c

```diff
@@ -297,15 +299,69 @@ static int set_pipeline_state(uint32_t id, uint32_t cmd)
 	if (ret < 0) {
 		tr_err(&ipc_tr, "ipc: comp %d trigger 0x%x failed %d", id, cmd, ret);
 		ret = IPC4_PIPELINE_STATE_NOT_SET;
-	} else if (ret > 0) {
-		msg_data.delayed_reply = true;
-		msg_data.delayed_error = IPC4_PIPELINE_STATE_NOT_SET;
+	} else if (ret == PPL_STATUS_SCHEDULED) {
+		*delayed = true;
 		ret = 0;
 	}
 
 	return ret;
 }
 
+static void ipc_compound_pre_start(int msg_id)
+{
+	/* ipc thread will wait for all scheduled ipc messages to be complete
+	 * Use a reference count to check status of these ipc messages.
+	 */
+	msg_data.delayed_reply++;
+}
+
+static void ipc_compound_post_start(uint32_t msg_id, int ret, bool delayed)
+{
+	if (ret) {
+		tr_err(&ipc_tr, "failed to process msg %d status %d", msg_id, ret);
+		msg_data.delayed_reply--;
+		return;
+	}
+
+	/* decrease counter if it is not scheduled by another thread */
+	if (!delayed)
+		msg_data.delayed_reply--;
+}
+
+static void ipc_compound_msg_done(uint32_t msg_id, int error)
+{
+	if (!msg_data.delayed_reply || !msg_reply.tx_size) {
+		tr_err(&ipc_tr, "unexpected delayed reply");
+		return;
+	}
+
+	msg_data.delayed_reply--;
+
+	/* error reported in delayed pipeline task */
+	if (error < 0) {
+		if (msg_id == SOF_IPC4_GLB_SET_PIPELINE_STATE)
+			msg_data.delayed_error = IPC4_PIPELINE_STATE_NOT_SET;
+	}
+}
+
+static int ipc_wait_for_compound_msg(void)
+{
+	int try_count = 10;
+	int ret = 0;
+
+	while (msg_data.delayed_reply) {
+		wait_delay(10000);
```

Comment:
This will need to be updated for Zephyr to ```yield(tmeout)```. Can you mark this as a TODO in the comments so we dont forget.

Reply:
sure